agents/smurf_stream_simulator/smurf_stream_simulator.py

```diff
@@ -153,17 +167,36 @@ def start_background_streamer(self, session, params=None):
                 time.sleep(1)
 
         # Teardown writer
-        writer.Close()
-        writer = None
+        self.writer.Close()
+        self.writer = None
 
         return True, "Finished streaming"
 
+    def _send_start_flowcontrol_frame(self):
+        """Send START flowcontrol frame."""
+        if self.writer is not None:
+            self.log.info("Sending START flowcontrol frame")
+            f = core.G3Frame(core.G3FrameType.none)
+            f['sostream_flowcontrol'] = FlowControl.START.value
+            self.writer.Process(f)
+
+    def _send_end_flowcontrol_frame(self):
+        """Send END flowcontrol frame."""
+        if self.writer is not None:
+            self.log.info("Sending END flowcontrol frame")
+            f = core.G3Frame(core.G3FrameType.none)
+            f['sostream_flowcontrol'] = FlowControl.END.value
+            self.writer.Process(f)
+
     def stop_background_streamer(self, session, params=None):
         """stop_background_streamer(params=None)
 
         Stop method associated with start_background_streamer process.
 
         """
+        if self.is_streaming:
+            self._send_end_flowcontrol_frame()
```

Comment:
Is this run in a separate thread than the "start_background_streamer" function? If so it seems its possible to send frames in the wrong order here. Like if you call `_send_end_flowcontrol_frame` before the Process call on line 161 then you'll get end_frame->dataframe->exit. Not sure if it matters so much but might be worth trying to keep the G3NetworkSender only writing from a single thread.

Reply:
Yeah, definitely it's already sending frames in the wrong order. One frame typically makes it after an "end" frame, as you pointed out. It worked quickly for testing prior to the meeting, but I agree it should be fixed.

Reply:
I've updated the simulator so that all write operations happen from the main streamer process, all the frame sending should happen from a single thread now.